classes/services/PKPPublicationService.inc.php

```diff
@@ -508,4 +557,95 @@ public function delete($publication) {
 
 		\HookRegistry::call('Publication::delete', [$publication]);
 	}
+
+	/**
+	 * Handle a publication setting for an uploaded file
+	 *
+	 * - Moves the temporary file to the public directory
+	 * - Resets the param value to what is expected to be stored in the db
+	 * - If a null value is passed, deletes any existing file
+	 *
+	 * This method is protected because all operations which edit publications should
+	 * go through the add and edit methods in order to ensure that
+	 * the appropriate hooks are fired.
+	 *
+	 * @param Publication $publication The publication being edited
+	 * @param Submission $submission The submission this publication is part of
+	 * @param mixed $value The param value to be saved. Contains the temporary
+	 *  file ID if a new file has been uploaded.
+	 * @param string $settingName The name of the setting to save, typically used
+	 *  in the filename.
+	 * @param integer $userId ID of the user who owns the temporary file
+	 * @param string $localeKey Optional. Pass if the setting is multilingual
+	 * @param boolean $isImage Optional. For image files which include alt text in value
+	 * @return string|array|null New param value or null on failure
+	 */
+	protected function _saveFileParam($publication, $submission, $value, $settingName, $userId, $localeKey = '', $isImage = false) {
+
+		// If the value is null, delete any existing unused file in the system
+		if (is_null($value)) {
+			$oldPublication = Services::get('publication')->get($publication->getId());
+			$oldValue = $oldPublication->getData($settingName, $localeKey);
+			$fileName = $oldValue['uploadName'];
+			if ($fileName) {
+				// File may be in use by other publications
+				$fileInUse = false;
+				foreach ((array) $submission->getData('publications') as $iPublication) {
```

Comment:
Stuff like this is where I think serialized data about files is not serving us well -- but not something to be fixed here.

Reply:
I totally agree. I would like to move towards a library/id approach. So that every file (at least every public file) would have an entry in tables like `file` and `file_settings`. And then there would be a table for linking files with objects, like `file_assoc` that has fields `file_id`, `assoc_id` and `assoc_type`.

That would make it much easier to associate a file (like an image) with all kinds of other objects on a case-by-case basis, without all the extra handling in the service classes. I'm thinking images for announcements, consistent generation of thumbnails, etc.